zebra-state/src/service/finalized_state/zebra_db/shielded.rs

```diff
@@ -329,6 +428,41 @@ impl DiskWriteBatch {
         self.prepare_history_batch(db, finalized, sapling_root, orchard_root, history_tree)
     }
 
+    /// Calculate the note commitment tree roots using parallel `rayon` threads.
+    fn note_commitment_roots_parallel(
+        note_commitment_trees: NoteCommitmentTrees,
+    ) -> (sprout::tree::Root, sapling::tree::Root, orchard::tree::Root) {
+        let NoteCommitmentTrees {
+            sprout,
+            sapling,
+            orchard,
+        } = note_commitment_trees;
+
+        let mut sprout_root = None;
+        let mut sapling_root = None;
+        let mut orchard_root = None;
+
+        rayon::in_place_scope_fifo(|scope| {
+            scope.spawn_fifo(|_scope| {
+                sprout_root = Some(sprout.root());
+            });
+
+            scope.spawn_fifo(|_scope| {
+                sapling_root = Some(sapling.root());
+            });
+
+            scope.spawn_fifo(|_scope| {
+                orchard_root = Some(orchard.root());
+            });
+        });
```

Comment:
This is how I run sprout, sapling, and orchard note commitment tree root calculations in parallel in the finalized state.

We could remove this code, and just calculate the roots after the note commitment tree update. (The root is cached.)

What do you think?

Reply:
We could run `sprout.root()` after updating the `sprout` tree, in the same `rayon` task that updates the `sprout` tree.

I'll make this change in a separate commit in PR #4802.